util/src/main/java/org/killbill/billing/util/dao/HistorySqlDao.java

```diff
@@ -29,11 +29,15 @@
 import org.skife.jdbi.v2.sqlobject.GetGeneratedKeys;
 import org.skife.jdbi.v2.sqlobject.SqlQuery;
 import org.skife.jdbi.v2.sqlobject.SqlUpdate;
+import org.skife.jdbi.v2.sqlobject.customizers.Define;
 
 public interface HistorySqlDao<M extends EntityModelDao<E>, E extends Entity> {
 
+    //  bypassMappingRegistryCache is to be used in the mapping registry to bypass cache. This is useful for generic classes since it will prevent to return previously cached class, that could result in wrong maps.
+    //  https://github.com/killbill/killbill-commons/blob/work-for-release-0.19.x/jdbi/src/main/java/org/skife/jdbi/v2/MappingRegistry.java#L67
     @SqlQuery
-    public List<EntityHistoryModelDao<M, E>> getHistoryForTargetRecordId(@Bind("targetRecordId") final long targetRecordId,
+    public List<EntityHistoryModelDao<M, E>> getHistoryForTargetRecordId(@Define("bypassMappingRegistryCache") final boolean bypassMappingRegistryCache,
```

Comment:
@sbrossie, After I completed the implementation to get the audit logs with history for the other objects and started testing, I noticed that the first call was good but the others calls where bad, since bring the last history entity but with new data (only matched fields). Digging out, I notice that all the mappers are cached via its class; in the case of the history the class is `EntityHistoryModelDao`, the difference is on the parameter passed into the class, but the map is stored and retrieved from the cache via the class, thus always fetching from cache the first one that was stored. That is why I introduced a flag to decide if store the mapper on the cache or not, and I am using the `@Define("bypassMappingRegistryCache")` to pass that flag to the register.